include/deal.II/multigrid/mg_constrained_dofs.h

```diff
@@ -146,13 +169,64 @@ MGConstrainedDoFs::initialize(const DoFHandler<dim,spacedim> &dof)
 {
   boundary_indices.clear();
   refinement_edge_indices.clear();
+  level_constraints.clear();
 
   const unsigned int nlevels = dof.get_triangulation().n_global_levels();
 
-  //At this point refinement_edge_indices is empty.
+  //At this point level_constraint and refinement_edge_indices are empty.
+  level_constraints.resize(nlevels);
   refinement_edge_indices.resize(nlevels);
   for (unsigned int l=0; l<nlevels; ++l)
-    refinement_edge_indices[l] = IndexSet(dof.n_dofs(l));
+    {
+      IndexSet relevant_dofs;
+      DoFTools::extract_locally_relevant_level_dofs(dof,l,relevant_dofs);
+      level_constraints[l].reinit(relevant_dofs);
+
+      // Loop through relevant cells and faces finding those which are periodic neighbors.
+      typename DoFHandler<dim,spacedim>::cell_iterator
+      cell = dof.begin(l),
+      endc = dof.end(l);
+      for (; cell!=endc; ++cell)
+        {
+          if (cell->level_subdomain_id() == numbers::artificial_subdomain_id)
+            continue;
+
+          for (unsigned int f = 0;
+               f < GeometryInfo<dim>::faces_per_cell;
+               ++f)
+            if (cell->has_periodic_neighbor(f))
+              {
+                if (cell->level_subdomain_id() ==  dof.get_triangulation().locally_owned_subdomain())
+                  {
+                    Assert(cell->periodic_neighbor(f)->level_subdomain_id() != numbers::artificial_subdomain_id,
+                           ExcMessage ("Periodic neighbor of a locally owned cell must either be owned or ghost."));
+                  }
+                else if (cell->periodic_neighbor(f)->level_subdomain_id() == numbers::artificial_subdomain_id)
+                  {
+                    continue;
+                  }
+
+                const unsigned int dofs_per_face = cell->face(f)->get_fe(0).dofs_per_face;
+                std::vector<types::global_dof_index> dofs_1(dofs_per_face);
+                std::vector<types::global_dof_index> dofs_2(dofs_per_face);
+
+                cell->periodic_neighbor(f)->face(cell->periodic_neighbor_face_no(f))->get_mg_dof_indices(l, dofs_1, 0);
+                cell->face(f)->get_mg_dof_indices(l, dofs_2, 0);
+                // Store periodicity information in the level constraint matrix
+                for (unsigned int i=0; i<dofs_per_face; ++i)
+                  if (level_constraints[l].can_store_line(dofs_2[i]) &&
+                      level_constraints[l].can_store_line(dofs_1[i]) &&
+                      !level_constraints[l].is_constrained(dofs_2[i]) &&
+                      !level_constraints[l].is_constrained(dofs_1[i]))
```

Comment:
Out of curiosity: Since you've reinitialized the `level_constraints` object at the beginning of working on this level, these conditions can only be true if you've added constraints for periodicity for them before. Maybe add a comment before the `if` that says something like "skip DoFs for which we've previously entered periodicity constraints already; this can happen, for example, for a vertex dof at a periodic boundary that we visit from more than one cell"